# Exploring Hacker News Posts

This project seeks to identify trends with Hacker News Posts. Hacker News is a site where user-submitted stories (a.k.a. posts) are voited and commented upon. We will be using an abridged version of the following dataset: https://www.kaggle.com/hacker-news/hacker-news-posts. The original version of the dataset has about 300,000 rows. We are using a dataset that is a random sample of 20,000 submissions. The 280,000 rows were removed because they did not receive any comments.

We will be looking at 2 main types of posts: Ask HN and Show HN. Ask HN posts are submitted when a user has a specific question they wish to ask the community. Show HN posts are submitted when a user would like to showcase a project, product, or some other interesting content to the community.

We will compare these two types of posts to determine:
1. Do Ask HN or Show HN receive more comments on average?
2. Do posts created at a certain time receive more comments on average?

## Importing the Data

In [1]:
# Put all imports here
from csv import reader
import datetime as dt

In [2]:
with open("hacker_news.csv") as hn_file:
    hn_content = reader(hn_file)
    hn = list(hn_content)
    
hn[:5]

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'],
 ['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '8/4/2016 11:52'],
 ['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20'],
 ['11919867',
  'Technology ventures: From Idea to Enterprise',
  'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429',
  '3',
  '1',
  'hswarna',
  '6/17/2016 0:01']]

## Removing the Header

We can see that in our hn object, the first item is the header. Let's remove the header and store in a variable so we can analyze our data.

In [3]:
headers = hn[0]
hn = hn[1:]

print("Header is:", headers)
print("\n")
print("First five rows are:", hn[:5])

Header is: ['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


First five rows are: [['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '

## Extracting only Ask HN and Show HN Posts

In [4]:
ask_posts = []
show_posts = []
other_posts = []

for row in hn:
    lower_title = row[1].lower()
    if lower_title.startswith('ask hn'):
        ask_posts.append(row)
    elif lower_title.startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)
        
print("Number of posts in ask_posts:", len(ask_posts))
print("Number of posts in show_posts:", len(show_posts))
print("Number of posts in other_posts:", len(other_posts))

Number of posts in ask_posts: 1744
Number of posts in show_posts: 1162
Number of posts in other_posts: 17194


## Average Number of Comments for Ask HN and Show HN Posts

In [5]:
total_ask_comments = 0
for row in ask_posts:
    total_ask_comments += float(row[4])
    
total_show_comments = 0
for row in show_posts:
    total_show_comments += float(row[4])
    
avg_ask_comments = total_ask_comments / len(ask_posts)
avg_show_comments = total_show_comments / len(show_posts)

print("Average number of comments for an Ask HN post:", avg_ask_comments)
print("Average number of comments for a Show HN post:", avg_show_comments)

Average number of comments for an Ask HN post: 14.038417431192661
Average number of comments for a Show HN post: 10.31669535283993


Based on the above findings, it looks like Ask HN posts receive almost 4 more comments on average. Ask HN posts on average receive 14.04 comments, while Show HN posts on average receive 10.32 comments.

Because of this, we will focus the rest of this project on Ask posts. Specifically, we will determine how the time of a post affects the number of comments it receives.

## Finding the Amount of Ask Posts and Comments by Hour Created

We will perform these two steps:
1. Calculate the number of ask posts created in each hour of the day, along with the number of comments received.
2. Calculate the average number of comments each ask post receives by hour created.

### Calculating Number of Ask Posts and Comments Per Hour

In [6]:
result_list = []
for post in ask_posts:
    created_at = dt.datetime.strptime(post[6], "%m/%d/%Y %H:%M")
    num_comments = int(post[4])
    result_list.append([created_at, num_comments])
    
# This will give us the number of ASK HN posts by hour.    
counts_by_hour = {}
# This will give us the number of comments on ASK HN posts by hour.
comments_by_hour = {}

for row in result_list:
    hour = row[0].hour
    if hour not in counts_by_hour:
        counts_by_hour[hour] = 1
    else:
        counts_by_hour[hour] += 1
    
    if hour not in comments_by_hour:
        comments_by_hour[hour] = row[1]
    else:
        comments_by_hour[hour] += row[1]
        
print("Posts By Hour:", counts_by_hour)
print("\n")
print("Comments By Hour:", comments_by_hour)

Posts By Hour: {9: 45, 13: 85, 10: 59, 14: 107, 16: 108, 23: 68, 12: 73, 17: 100, 15: 116, 21: 109, 20: 80, 2: 58, 18: 109, 3: 54, 5: 46, 19: 110, 1: 60, 22: 71, 8: 48, 4: 47, 0: 55, 6: 44, 7: 34, 11: 58}


Comments By Hour: {9: 251, 13: 1253, 10: 793, 14: 1416, 16: 1814, 23: 543, 12: 687, 17: 1146, 15: 4477, 21: 1745, 20: 1722, 2: 1381, 18: 1439, 3: 421, 5: 464, 19: 1188, 1: 683, 22: 479, 8: 492, 4: 337, 0: 447, 6: 397, 7: 267, 11: 641}


We now have two dictionaries. counts_by_hour gives us the number of ask posts in each hour. comments_by_hour gives us the number of comments on ask posts in each hour. We will now create a list of lists, where each element is a 2-item list, with the first item being the hour, and the second item being the average number of comments per post.

In [7]:
avg_by_hour = []
for hour in counts_by_hour:
    num_posts = counts_by_hour[hour]
    num_comments = comments_by_hour[hour]
    avg_comments_per_post = num_comments / num_posts
    avg_by_hour.append([hour, avg_comments_per_post])
    
print(avg_by_hour)

[[9, 5.5777777777777775], [13, 14.741176470588234], [10, 13.440677966101696], [14, 13.233644859813085], [16, 16.796296296296298], [23, 7.985294117647059], [12, 9.41095890410959], [17, 11.46], [15, 38.5948275862069], [21, 16.009174311926607], [20, 21.525], [2, 23.810344827586206], [18, 13.20183486238532], [3, 7.796296296296297], [5, 10.08695652173913], [19, 10.8], [1, 11.383333333333333], [22, 6.746478873239437], [8, 10.25], [4, 7.170212765957447], [0, 8.127272727272727], [6, 9.022727272727273], [7, 7.852941176470588], [11, 11.051724137931034]]


## Improving Readability of Results

We have a list that contains the average number of comments per Ask HN post for each hour. However, it is quite hard to read and understand. Let's finish by sorting the list of lists and printing the five highest values in a readable format.

Start by swapping the order of the elements so we can call the sorted function and sort by avg number of comments.

In [8]:
swap_avg_by_hour = []
for element in avg_by_hour:
    swap_avg_by_hour.append([element[1], element[0]])
    
print(swap_avg_by_hour)

[[5.5777777777777775, 9], [14.741176470588234, 13], [13.440677966101696, 10], [13.233644859813085, 14], [16.796296296296298, 16], [7.985294117647059, 23], [9.41095890410959, 12], [11.46, 17], [38.5948275862069, 15], [16.009174311926607, 21], [21.525, 20], [23.810344827586206, 2], [13.20183486238532, 18], [7.796296296296297, 3], [10.08695652173913, 5], [10.8, 19], [11.383333333333333, 1], [6.746478873239437, 22], [10.25, 8], [7.170212765957447, 4], [8.127272727272727, 0], [9.022727272727273, 6], [7.852941176470588, 7], [11.051724137931034, 11]]


In [9]:
sorted_swap = sorted(swap_avg_by_hour, reverse=True)
sorted_swap

[[38.5948275862069, 15],
 [23.810344827586206, 2],
 [21.525, 20],
 [16.796296296296298, 16],
 [16.009174311926607, 21],
 [14.741176470588234, 13],
 [13.440677966101696, 10],
 [13.233644859813085, 14],
 [13.20183486238532, 18],
 [11.46, 17],
 [11.383333333333333, 1],
 [11.051724137931034, 11],
 [10.8, 19],
 [10.25, 8],
 [10.08695652173913, 5],
 [9.41095890410959, 12],
 [9.022727272727273, 6],
 [8.127272727272727, 0],
 [7.985294117647059, 23],
 [7.852941176470588, 7],
 [7.796296296296297, 3],
 [7.170212765957447, 4],
 [6.746478873239437, 22],
 [5.5777777777777775, 9]]

In [10]:
print("Top 5 Hours for Ask Posts Comments")

Top 5 Hours for Ask Posts Comments


In [11]:
for element in sorted_swap[:5]:
    hour = str(element[1])
    hour_dt = dt.datetime.strptime(hour, "%H")
    formatted_hour = dt.datetime.strftime(hour_dt, "%H:%M")
    formatted_string = "{formatted_hour}: {avg_comments:.2f} average comments per post".format(
        formatted_hour=formatted_hour, avg_comments=element[0])
    print(formatted_string)

15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post


## Findings

Based on the above data, Ask HN posts that are posted during 15:00, 02:00, 20:00, 16:00, and 21:00 EST have the highest chance of receiving comments.

Let's explore the data a bit further.

## Do Show or Ask Posts Receive More Points On Average?

## Do Posts Created At Certain Times Receive More Points?

## Comparing Avg Comments And Points to Other (non-Ask, non-Show) Posts